### DISCUSSION TITLE
What happens with debt when a stream is canceled?

### DISCUSSION THREAD

**Discussion by razgraf** [sablier-labs/flow#58] [Q&A] [answered]
Say I have a stream of 1000$/month, I keep it open for 3 months but only fund it with 1000$. At the end of the period, I have 2000$ in debt. 

What happens when I cancel/stop the payment?
- should the sender be allowed to cancel only if they pay the debt? (might be linked to #57, as they could ask the recipient to void it first if they don't want to pay)
- should I be able to cancel/stop/pause it without paying the debt?

Follow-up questions: based on the answers for the above, should we have utilities like "payDebtAndCancelStream" to be able do do everything in one tx?

CC: @smol-ninja @andreivladbrg

**Comment by smol-ninja** [comment]
> What happens when I cancel/stop the payment?

In the current design, the sender will be able to cancel. The consequence of not allowing them to cancel when there is a debt is that they can just ignore it. There is an assumption that there will a mutual understanding between the sender and the recipient. One party can, of course, rug another party by not fulfilling his debt obligations.

> Follow-up questions: based on the answers for the above, should we have utilities like "payDebtAndCancelStream" to be able to do everything in one tx?

Yeah, we can. I like the idea.

**Reply by razgraf** [reply]
Would be great if you guys could offer a summary of `pause` and `void` here so I can mark that answer as correct for future reference 🙏 Or link to an existing discussion that's related if we have any. Thank you!

**Reply by smol-ninja** [reply, marked as answer]
You can refer to the "Table" in [this discussion](https://github.com/sablier-labs/v2-open-ended/discussions/86).

To summarize:
- `pause()` will pause the stream by setting `ratePerSecond` to zero. It will also store debt accumulated until then. It can only be called by the sender.
- `void()` will also pause the stream but it will set `debt` to zero. It can only be called by the recipient.

cc @andreivladbrg.

**Comment by andreivladbrg** [comment]
> Follow-up questions: based on the answers for the above, should we have utilities like "payDebtAndCancelStream" to be able to do everything in one tx?

great idea @razgraf 🫡

https://github.com/sablier-labs/v2-open-ended/issues/59

**Comment by smol-ninja** [comment]
@razgraf wanted to hear your opinion on this approach? Should it continue to remain the same or should it not reset debt to 0?

**Reply by razgraf** [reply]
Resetting debt to zero is fine.

By default, we'll suggest the user to call `payDebtAndCancel` when they cancel as this is the "fair" assumed behavior (you want to pay what is owed). If they choose the forced `cancel` option, we can handle things on the UI side and highlight when a stream has been stopped with outstanding debt.

Do you guys think this would be a good way to implement the cancel flow?

**Reply by andreivladbrg** [reply]
> Do you guys think this would be a good way to implement the cancel flow?

Personally, I find the current flow satisfactory. However, @smol-ninja has suggested adding a new `debt` variable: https://github.com/sablier-labs/v2-open-ended/discussions/64.

Since we will have `payDebtAndCancel`, I think that would suffice, and adding a new variable wouldn't be needed.

**Reply by razgraf** [reply]
That's an interesting approach too 🤔 
 
- would simplify the list of methods
- would hint at the importance of `debt` being paid at time of cancel (any cancel)
- would however make separation of concerns harder, as it would make "cancel" have more than one purpose

I know @PaulRBerg is usually promoting simplicity in the responsibility of methods. Because of that I'm leaning towards having two separate "core" method (`payDebt`, `cancel`) and an aggregated "peripheral" one (`payDebtAndCancel`).

At the same time... cannot understate the beauty of a self-explanatory API (one method = "hey, when you cancel, always consider how the debt may affect your agreements").

Sorry for potentially not being helpful here :D

**Reply by andreivladbrg** [reply]
> Because of that I'm leaning towards having two separate "core" method (payDebt, cancel) and an aggregated "peripheral" one (payDebtAndCancel).

It may not be possible to have `payDebtAndCancel` in a separate contract, as the sender needs to be the caller in OE contract

**Reply by razgraf** [reply]
Ah yes the sender for cancel, you're right.

**Comment by smol-ninja** [comment]
>  the same time... cannot understate the beauty of a self-explanatory API

I like the approach of designing core APIs such that each of them does just one job. i.e. `_adjustRatePerSecond()` only adjusts the rate per second without transferring any asset, `_cancel()` only cancels the stream without making any transfers.

It can make them more composable, easier to implement a new API without re-deploying the core contract. Now to achieve complex interactions such as cancel and refund or restart and deposit, there could be 3 ways (or more):
- **Separate periphery contracts:**  It comes with problem that @andreivladbrg has pointed out. This can be taken care of with either `delegatecall()` but because it is a separate contract, it increases the complexity compared to the advantage it offers. 

- **Implement periphery related functions in the core contracts:** this is how we are doing it right now but it limits the number of batch functions we can write. We will still have to deploy a new contract if there is a requirement for a new API.

- **Allow UI to execute a combination of functions directly on the core contract**

And I think the 3rd point gives us a really good food for thought. I have [shared it before](https://github.com/sablier-labs/v2-open-ended/discussions/55#discussioncomment-9285605) and it was discarded for the following reason:

> I would definitely like to have named parameters and structs instead of encoded bytes for the methods as it makes debugging easier.
Bytes are probably more efficient and scalable (that batch method acts a bit like execute(bytes) in safes) but DX-wise, not sure we need it right now.

But I would like to pick it again because I think I agreed easily with what @razgraf pointed out earlier.

```solidity
function batch(bytes[] calldata calls, bool revertOnFail) external payable {
    for (uint256 i = 0; i < calls.length; i++) {
        (bool success, bytes memory result) = address(this).delegatecall(calls[i]);
        if (!success && revertOnFail) {
            _getRevertMsg(result);
        }
    }
}
```

> I would definitely like to have named parameters and structs instead of encoded bytes for the methods as it makes debugging easier.

Can you please explain why would you have to deal with raw bytes directly? In the UI, isn't it possible to use the function named parameters to construct the calls array?

> Bytes are probably more efficient and scalable (that batch method acts a bit like execute(bytes) in safes) but DX-wise, not sure we need it right now.

The above function does not make it efficient at all. And yes it acts similarly to `execute(bytes)` but it is a much simpler version. If you look at it, it just calls all the functions mentioned in the call parameter to itself.

The above function is so useful that it wouldn't even require us to deploy periphery contract (until they offer some benefits) and we can absolutely expose any API through the user interface.

I would like to hear criticism against this approach again. Tagging @sablier-labs/solidity for comments.

FWIW Uniswap also has [a similar function](https://etherscan.io/address/0x3fc91a3afd70395cd496c647d5a6cc9d4b2b7fad#code#F1#L28).

**Reply by razgraf** [reply]
I will share my thoughts on this tomorrow with fresh eyes - wanna digest the pros/cons with a clear mind. 

The one thing I wanted to mention right now:

> Can you please explain why would you have to deal with raw bytes directly?

I was suggesting difficulties here not for our UI, but for interacting with the contracts on Etherscan, Tenderly and similar UIs, with the purpose of debugging. Having bytes as input data requires us to depend on an external script to encode it and another one to decode it vs. simply adding the parameters one by one.

**Reply by razgraf** [reply]
I've read this again. I like this `batch` method.

My hunch is it wouldn't work for state-heavy periphery (e.g. cancel-and-create) but I guess we don't use/have those at the moment. 

I still think the difficulties highlighted by my reply above are valid, so I'd love to hear your take on this.

Also, given this is so general purpose are there any attack vectors it could promote?

**Reply by andreivladbrg** [reply]
> given this is so general purpose are there any attack vectors it could promote?

we will have tests for it, but there shouldn’t be any attack vectors if the underlying functions (`create`, `cancel,` etc.) which are batch called, don’t have issues

**Reply by smol-ninja** [reply]
> it wouldn't work for state-heavy periphery (e.g. cancel-and-create)

I believe it can also work for `cancel-and-create` as long as `cancel` and `create` work on their own. The batch function would simply execute them sequentially in a single transaction, where `create` follows `cancel`.

Note that `batch` function can only be called with public functions, thus it does not make an efficient execution. So how about we still implement the key "important" functions (the most fundamental ones) in the core contract itself such as `createAndDeposit` which would make internal calls and thus, will be more optimized over `batch` call.

> I was suggesting difficulties here not for our UI, but for interacting with the contracts on Etherscan

Agree, but even our lockup contracts are not easy to interact with on Etherscan due to the use of `params` struct in the `create` functions. Right?

> Tenderly and similar UIs

Since the `batch` function still calls publicly exposed functions of the core contracts, Tenderly should not have problems with debugging. However, I agree this is a valid concern to review transactions by simply looking at the Etherscan explorer.

> Having bytes as input data requires us to depend on an external script to encode it and another one to decode

True, but I wouldn't say "external script". Constructing input data would require access to open ended interface. As I mentioned above, this is not very different than having `params` struct in the lockup contracts. 

Similarly, anyone can use an open-ended interface to encode and decode data required for batch functions.

An example:

```solidity
bytes[] memory calls = new bytes[](2);
calls[0] = abi.encodeWithSelector(ISablierOpenEnded.create.selector, sender, recipient, ratePerSecond, asset);
calls[1] = abi.encodeWithSelector(ISablierOpenEnded.deposit.selector, OpenEnded.nextStreamId(), amount);
openEnded.batch(calls, true);
```

> Also, given this is so general purpose are there any attack vectors it could promote?

Since it's a delegate call to itself and `batch` can only be applied to public functions, no additional risks are introduced by it.

**Reply by andreivladbrg** [reply]
question @smol-ninja: to make this work for us, we would need to remove the `noDelegateCall` modifier.

If we are going to do this, why not completely separate the `batch` logic from this contract and use a `Multicall` contract instead?

**Reply by smol-ninja** [reply]
No, there is no need to remove the `noDelegateCall` modifier. The delegate call is made from the contract itself and not from an outside proxy contract. 

```solidity
function _preventDelegateCall() private view {
    if (address(this) != ORIGINAL) {
        revert Errors.DelegateCall();
    }
}
```

In the above function, `address (this) == ORIGINAL` always.

**Reply by andreivladbrg** [reply]
sheesh, you are right

**Reply by razgraf** [reply]
>  our lockup contracts are not easy to interact with on Etherscan due to the use of params struct

Totally, not the easiest to interact with but still better than `bytes`.

> True, but I wouldn't say "external script". Constructing input data would require access to open ended interface. As I mentioned above, this is not very different than having params struct in the lockup contracts.

But it is. The prerequisite cost is much higher, while the time-to-result is longer.
- with params I only need to type a JSON object at max (when Etherscan/Tenderly doesn't simplify it for me due to `params`), I can do it manually directly in the input form fields
- with bytes I need a terminal or a script or an app that takes in the ABIs and converts everything to bytes

You are right to assume that for developers (physically close to an IDE) this shouldn't be a pain, but for other people going for a quick test, looking to move fast or trying to interact with the contract through Etherscan/Tenderly it becomes a pain, right?

> How about we still implement the key "important" functions (the most fundamental ones) in the core contract itself such as createAndDeposit which would make internal calls and thus, will be more optimized over batch call.

I like this idea, although wouldn't it introduce a bit of fragmentation of features?

**Reply by smol-ninja** [reply]
> looking to move fast or trying to interact with the contract through Etherscan/Tenderly it becomes a pain, right?

I agree with you. I just had a look at etherscan, and it simplifies the struct parameter very neatly. So yeah it would be a pain to use `batch` from etherscan. But maybe not so much. Etherscan also shows the function selector next to the function name. So with sufficient knowledge, one can easily construct input data for the `batch` function without requiring an external script.

<img width="318" alt="Screenshot 2024-05-15 at 22 08 57" src="https://github.com/sablier-labs/v2-open-ended/assets/6676622/77af1739-2581-44d9-8721-e9fd9b976a39">


Wdyt?

> I like this idea, although wouldn't it introduce a bit of fragmentation of features?

Can you please add more explanation on how it would create fragmentation?

**Reply by razgraf** [reply]
> one can easily construct input data for the batch function without requiring an external script

True, but my point is: they still have to [run](https://viem.sh/docs/contract/encodeFunctionData#encodefunctiondata) `encodeFunctionData({abi, method, variables, ...})` or similar to convert that to bytes, don't they?

> fragmentation

Sorry my bad, I misread your solution as having those key important functions be implemented in a periphery contract still. You meant have everything in the core: simple methods, key compounded methods and the batch, right?

---

Note: I'm not against the idea of dropping the periphery if in the end it just makes sense to, I'm just trying to poke holes so we don't miss something (including DX).

**Reply by smol-ninja** [reply]
> True, but my point is: they still have to [run](https://viem.sh/docs/contract/encodeFunctionData#encodefunctiondata) encodeFunctionData({abi, method, variables, ...}) or similar to convert that to bytes, don't they?

I agree with you. They will.

> Sorry my bad, I misread your solution as having those key important functions be implemented in a periphery contract still. You meant have everything in the core: simple methods, key compounded methods and the batch, right?

Yes thats what I meant.

> Note: I'm not against the idea of dropping the periphery if in the end it just makes sense to, I'm just trying to poke holes so we don't miss something (including DX).

Yes I agree, its important to have this discussion before we make any conclusion.

@andreivladbrg do you anticipate any security concern using `batch` function other that what we have already discussed?

**Reply by andreivladbrg** [reply]
>@andreivladbrg do you anticipate any security concern using batch function other that what we have already discussed?

nope, there isn't anything else i can think of atm

**Reply by razgraf** [reply]
@smol-ninja would you mind moving the idea of this batch function into a dedicated discussion? 

If you think it would improve our contract design, let's give it a shot - we'll implement it, and after the first sprint evaluate the DX behind it (after we bake it into the app too). If it's clearly annoying, we can simply go back to a periphery design or go over the aggregated-core-methods idea again (having everything in the core).

**Reply by smol-ninja** [reply]
Agree. Started here: https://github.com/sablier-labs/v2-open-ended/discussions/88

**Comment by PaulRBerg** [comment]
To avoid having to read all comments here, could you guys confirm that this discussion has effectively been deprecated in favor of @smol-ninja's proposal to [carry forward the debt](https://github.com/sablier-labs/v2-open-ended/discussions/86)?

If yes, we should close this as outdated.

**Reply by smol-ninja** [reply]
Yes we can.
